### DISCUSSION TITLE
HPN-SSH 18.3.0 Released

### DISCUSSION THREAD

**Discussion by rapier1** [rapier1/openssh-portable#65] [Announcements]
HPN-SSH 18.3.0 incorporates the most recent OpenSSH release, 9.6p1. This version does not introduce any new functionality. However, this release addresses a potential bug that may be encountered when an OpenSSH install is sending data to an HPN-SSH install. In some circumstances, especially at higher RTTs, it is possible that the transfer will enter into a pathological state typified by greatly reduced performance and excessive CPU usage on the sender side. We believe that this might be caused by poor interactions in OpenSSH when large receive windows are advertised to the client. This has been addressed by reducing the maximum size of the window advertised to OpenSSH to 15MB. This is a non-optimal solution especially in high RTT environments. We will be continuing to explore the causes and work towards creating a more performant solution for these situations. 

Packages are available from:
ubuntu: sudo add-apt-repository ppa:rapier1/hpnssh
fedora: sudo dnf copr enable rapier1/hpnssh
Debian 11: https://download.opensuse.org/repositories/home:/rapier1/Debian_11/
Debian 12: https://download.opensuse.org/repositories/home:/rapier1/Debian_12/

**Comment by IceCodeNew** [comment]
Here are the tips for the Debian users:

# 1. Manually add the package source
```shell
. /etc/os-release
suites="Debian_${VERSION_ID}"

sudo mkdir -p --mode=0755 /usr/share/keyrings
out=$(mktemp -u)
if curl -fsSL "https://download.opensuse.org/repositories/home:/rapier1/${suites}/Release.key" \
    | gpg --dearmor | sudo tee "$out" > /dev/null; then
    sudo \mv -f "$out" /usr/share/keyrings/rapier1-hpnssh.gpg
fi

cat << EOF | sudo tee /etc/apt/sources.list.d/hpnssh.sources
Types: deb
URIs: https://download.opensuse.org/repositories/home:/rapier1/${suites}/
Suites: /
Signed-By: /usr/share/keyrings/rapier1-hpnssh.gpg
EOF
```

# 2. Install the hpnssh-server package
```shell
sudo chmod 644 /usr/share/keyrings/rapier1-hpnssh.gpg /etc/apt/sources.list.d/hpnssh.sources \
    && sudo apt-get update -qq \
    && sudo apt-get install -y hpnssh-client hpnssh-server hpnssh-sftp-server

sudo systemctl enable --now hpnssh.service
systemctl is-enabled hpnssh.service
```

**Reply by justinclift** [reply]
Thanks.  Just tried this out on a brand new "minimal" install (eg no GUI, just base packages), as there are no other instructions for installing hpn-ssh on Debian anywhere, and the official `add-apt-repository` tool seems to hate OBS repositories (very frustrating). :wink:

Btw, the above instructions fail on a minimal install due to some common dependencies not being installed (yet) on minimal systems:

```
-bash: sudo: command not found
-bash: gpg: command not found
-bash: curl: command not found
-bash: sudo: command not found
-bash: sudo: command not found
```

Once they're installed though (`apt install -y gpg sudo curl`), the above does set the repository up properly, and `hpn-ssh` can be installed afterwards. :smile:

@rapier1 Is it feasible to have the above instructions added somewhere more prominent than here?  Maybe in an official README or something?

Without them, people (like me) will probably be trying to use the `apt-add-repository` cli tool, and that tool frankly doesn't want to work with the hpn-ssh repository structure.  It just fails outright (below), unlike the instructions above:

```
# apt-add-repository https://download.opensuse.org/repositories/home:/rapier1/Debian_12/
Repository: 'deb https://download.opensuse.org/repositories/home:/rapier1/Debian_12/ bookworm main'
Description:
Archive for codename: bookworm components: main
More info: https://download.opensuse.org/repositories/home:/rapier1/Debian_12/
Adding repository.
Press [ENTER] to continue or Ctrl-c to cancel.
Adding deb entry to /etc/apt/sources.list.d/archive_uri-https_download_opensuse_org_repositories_home_rapier1_debian_12_-bookworm.list
Adding disabled deb-src entry to /etc/apt/sources.list.d/archive_uri-https_download_opensuse_org_repositories_home_rapier1_debian_12_-bookworm.list
Hit:1 http://deb.debian.org/debian bookworm InRelease
[...]
Ign:6 https://download.opensuse.org/repositories/home:/rapier1/Debian_12 bookworm InRelease                                 
Get:7 https://download.opensuse.org/repositories/home:/rapier1/Debian_12  InRelease [1,865 B]
Err:8 https://download.opensuse.org/repositories/home:/rapier1/Debian_12 bookworm Release
  404  Not Found [IP: 195.135.223.226 443]
Reading package lists... Done
E: The repository 'https://download.opensuse.org/repositories/home:/rapier1/Debian_12 bookworm Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
```

That's a fail, and it won't find the hpn-ssh packages.

**Comment by IceCodeNew** [comment]
**FIXED**

~~This is what you would get if you tried to install the package `hpnssh-server` while the package `openssh-server` still exists:~~

```
# sudo apt install hpnssh-server hpnssh-client

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
hpnssh-client is already the newest version (1:9.6p1-hpn18.3.0-1ubuntu3).
Suggested packages:
  molly-guard monkeysphere hpnssh-askpass ufw
The following NEW packages will be installed:
  hpnssh-server
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
2 not fully installed or removed.
Need to get 0 B/498 kB of archives.
After this operation, 1932 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
Preconfiguring packages ...
(Reading database ... 48399 files and directories currently installed.)
Preparing to unpack .../hpnssh-server_1%3a9.6p1-hpn18.3.0-1ubuntu3_amd64.deb ...
Unpacking hpnssh-server (1:9.6p1-hpn18.3.0-1ubuntu3) ...
dpkg: error processing archive /var/cache/apt/archives/hpnssh-server_1%3a9.6p1-hpn18.3.0-1ubuntu3_amd64.deb (--unpack):
 trying to overwrite '/etc/sv/ssh/.meta/installed', which is also in package openssh-server 1:9.2p1-2+deb12u2
Errors were encountered while processing:
 /var/cache/apt/archives/hpnssh-server_1%3a9.6p1-hpn18.3.0-1ubuntu3_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by rapier1** [comment]
Crap. I'll get working on this. I gotta tell you, writing the code has been less of a pain in the butt than building the packages.

**Comment by rapier1** [comment]
I think I got it resolved. There should be new binaries in the debian repo shortly.

**Comment by justinclift** [comment]
If anyone's interested, I've thrown together my notes on installing hpn-ssh for Proxmox 8.x here:

https://github.com/rapier1/hpn-ssh/wiki/Installation-on-Proxmox-8.x

It's based heavily upon the approach by @IceCodeNew [above](https://github.com/rapier1/hpn-ssh/discussions/65#discussioncomment-8359796), with customisation for how Proxmox is used.

Haven't linked it into the index page of the wiki anywhere, as it looks like @rapier1 is putting proper documentation in the wiki and this might not really fit. :wink:

Feel free to edit/move/etc it as needed of course. :smile:
